libs/libc/gdbstub/lib_gdbstub.c

```diff
@@ -0,0 +1,1559 @@
+/****************************************************************************
+ * libs/libc/gdbstub/lib_gdbstub.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <ctype.h>
+#include <stdint.h>
+#include <string.h>
+#include <stdio.h>
+#include <syslog.h>
+#include <sys/param.h>
+#include <sys/types.h>
+
+#include <nuttx/sched.h>
+#include <nuttx/gdbstub.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#ifdef CONFIG_LIB_GDBSTUB_DEBUG
+#  define GDB_DEBUG(...) syslog(LOG_DEBUG, ##__VA_ARGS__)
+#  define GDB_ASSERT() __assert(__FILE__, __LINE__, 0)
+#else
+#  define GDB_DEBUG(...)
+#  define GDB_ASSERT() do {} while (0)
+#endif
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+struct gdb_state_s
+{
+  gdb_send_func_t send;                   /* Send buffer to gdb */
+  gdb_recv_func_t recv;                   /* Recv buffer from gdb */
+  FAR void *priv;                         /* Private data for transport */
+  gdb_monitor_func_t monitor;             /* Monitor will be called when gdb
+                                           * receive a monitor command
+                                           */
+  pid_t pid;                              /* Gdb current thread */
+  FAR char *pkt_next;                     /* Pointer to next byte in packet */
+  char pkt_buf[1024];                     /* Packet buffer */
+  size_t pkt_len;                         /* Packet send and receive length */
+  uint8_t running_regs[XCPTCONTEXT_SIZE]; /* Registers of running thread */
+  size_t size;                            /* Size of registers */
+  uintptr_t registers[0];                 /* Registers of other threads */
```

Comment:
Better to use `uintptr_t registers[1];` as I'm not sure if `a[0];` is C89 compatible